### PR TITLE
feat(trade): move import token and approval modals to TradeWidget

### LINKED ISSUE #2428: Move ImportTokenModal and TradeApproveWidget to TradeWidget
`ImportTokenModal` and `TradeApproveWidget` are currently integrated in `SwapModals` and `LimitOrdersWidget`.
Since we have a generic `TradeWidget` component, they should be there.

TODO: https://github.com/cowprotocol/cowswap/blob/feature/twap-orders-5/src/cow-react/modules/trade/containers/TradeWidget/index.tsx#L51

### PR DESCRIPTION
**Pull request by shoom3301**
# Summary

Fixes #2428

Since `<TradeWidget>` is a single point for all trade widgets, it makes sense to have there common modals like: 
 - `ImportTokenModal`
 - `ZeroApprovalModal`
 - `TradeApproveModal`

In this PR all those modals were moved to `TradeWidget`.  
Besides that, `swapDerivedStateAtom` was added and used in `useDerivedTradeState()` in order to make `useShouldZeroApproveSwap` hook universal for any trade widget.
Also, `sellAmountMaxSpend` parameter is added to `TradeDerivedState` for the same purposes ([commit](https://github.com/cowprotocol/cowswap/pull/2547/commits/9ace6a1f235b4151b78cb371c96f9f25e02980b8)).

  # To Test

1. Zero approve for Swap/Limit orders
2. Regular approve for Swap/Limit orders
3. Import token for Swap/Limit orders

### CODE REVIEW

File: src/modules/swap/containers/SwapWidget/index.tsx [hunk 1]
```diff
@@ -44,12 +44,14 @@ import SettingsTab from 'legacy/components/Settings'
 import { SwapButtonState } from 'modules/swap/helpers/getSwapButtonState'
 import { useIsEthFlow } from 'modules/swap/hooks/useIsEthFlow'
 import { useShouldZeroApproveSwap } from 'common/hooks/useShouldZeroApproveSwap'
+import { useFillSwapDerivedState } from '../../state/useSwapDerivedState'
```
Comment (nenadV91) [inline src/modules/swap/containers/SwapWidget/index.tsx:47]: Use absolute paths?

**Review by nenadV91 — APPROVED** [review]

File: src/modules/swap/containers/SwapWidget/index.tsx [hunk 2]
```diff
@@ -65,7 +67,7 @@ export function SwapWidget() {
   const { independentField, recipient } = swapState
   const showRecipientControls = useShowRecipientControls(recipient)
   const isEthFlow = useIsEthFlow()
-  const shouldZeroApprove = useShouldZeroApproveSwap()
+  const shouldZeroApprove = useShouldZeroApprove(sellAmountMaxSpend)
```
Comment (shoom3301) [inline src/modules/swap/containers/SwapWidget/index.tsx:70]: @tukantje please, pay attention here

File: src/legacy/state/swap/hooks.tsx [hunk 3]
```diff
@@ -83,6 +83,8 @@ interface DerivedSwapInfo {
   currenciesIds: { [field in Field]?: string | null }
   currencyBalances: { [field in Field]?: CurrencyAmount<Currency> }
   parsedAmount: CurrencyAmount<Currency> | undefined
+  // TODO: remove duplications of the value (v2Trade?.maximumAmountIn(allowedSlippage))
+  sellAmountMaxSpend: CurrencyAmount<Currency> | null
```
Comment (anxolin) [inline src/legacy/state/swap/hooks.tsx:87]: Are we sure we want this and not the allowance?

File: src/modules/trade/types/TradeDerivedState.ts [hunk 4]
```diff
@@ -6,6 +6,18 @@ export interface TradeDerivedState {
...
+   * Sometimes we need a maximum amount to sell
+   *
+   * How it works:
+   * You want to buy exactly 2 WETH for <some amount> of COW.
+   * In this case, you will input 2 WETH in the buy input
+   * and the sell input will be automatically filled by the quote API response, for example with the value 552 COW.
+   * Since the market is liquid, there might be a slippage, let's assume it's 2%.
+   * So, the final sell amount will be 552 COW + 2% = 563.04
+   */
+  readonly sellAmountMaxSpend: CurrencyAmount<Currency> | null
```
Comment (alfetopito) [inline src/modules/trade/types/TradeDerivedState.ts:20]: What about naming it `slippageAdjustedSellAmount`?

You can also add to the docstring info that for Sell orders and for any type of Limit orders, this is always equal to sell amount

File: src/modules/swap/hooks/useSwapButtonContext.ts [hunk 5]
```diff
@@ -64,13 +64,14 @@ export function useSwapButtonContext(input: SwapButtonInput): SwapButtonsContext
   const { isNativeIn, isWrappedOut, wrappedToken } = useDetectNativeToken()
   const isNativeInSwap = isNativeIn && !isWrappedOut
 
-  const nativeInput = trade?.maximumAmountIn(allowedSlippage)
-  const wrapUnwrapAmount = isNativeInSwap ? (nativeInput || parsedAmount)?.wrapped : nativeInput || parsedAmount
+  const wrapUnwrapAmount = isNativeInSwap
+    ? (sellAmountMaxSpend || parsedAmount)?.wrapped
+    : sellAmountMaxSpend || parsedAmount
```
Comment (alfetopito) [inline src/modules/swap/hooks/useSwapButtonContext.ts:69]: I know not part of this change, but can you create a variable for the amount?

```suggestion
  const amount = sellAmountMaxSpend || parsedAmount
  const wrapUnwrapAmount = isNativeInSwap ? (amount)?.wrapped : amount
```

**Comment by elena-zh** [comment]
Works fine to me. 
The only one thing that I noticed that warnings are clued on the Limit orders page. However, not sure if it is related to the current PR (I saw then in #2547 as well)
paddings

File: src/legacy/state/swap/hooks.tsx (hunk 3 above)
Comment (shoom3301) [reply src/legacy/state/swap/hooks.tsx:87]: Sorry, I don't understand. Why do wee need allowance here?

File: src/modules/swap/hooks/useSwapButtonContext.ts (hunk 5 above)
Comment (shoom3301) [reply src/modules/swap/hooks/useSwapButtonContext.ts:69]: Thank you! This looks better

File: src/modules/trade/types/TradeDerivedState.ts (hunk 4 above)
Comment (shoom3301) [reply src/modules/trade/types/TradeDerivedState.ts:20]: Thank you! Renamed

File: src/modules/swap/containers/SwapWidget/index.tsx (hunk 2 above)
Comment (shoom3301) [reply src/modules/swap/containers/SwapWidget/index.tsx]: Yes, the goal of these changes is the unification of getting an amount to approve - `slippageAdjustedSellAmount`.
Now this value is encapsulated in `TradeDerivedState` (each trade widget implements this interface).
Also, `<ZeroApprovalModal />` is integrated in `<TradeWidgetModals/>` which in turn provides the zero approval modal for each trade widget out of the box.

**Comment by shoom3301** [comment]
@elena-zh I can't reproduce this case with 3 warnings :( But I guess it's not related to the current changes. Could you compare with develop please?

**Comment by elena-zh** [comment]
> @elena-zh I can't reproduce this case with 3 warnings :( But I guess it's not related to the current changes. Could you compare with develop please?

The issue is also reproducible on Develop. However, Develop is not updated with the latest version (1.38.1).

**Review by alfetopito — APPROVED** [review]
